plusdeck/dbus/client.py

```diff
@@ -0,0 +1,335 @@
+from plusdeck.client import State
+from plusdeck.config import Config
+from plusdeck.dbus.interface import DBUS_NAME, DbusInterface
+
+logger = logging.getLogger(__name__)
+
+
+class DbusClient(DbusInterface):
+    """
+    Plus Deck 2C dbus client.
+    """
+
+    def __init__(self: Self) -> None:
+        client = Mock(name="client", side_effect=NotImplementedError("client"))
+        super().__init__("", client)
+        cast(Any, self)._proxify(DBUS_NAME, "/")
+
+
+@dataclass
+class Obj:
+    """
+    The main click context object. Includes a dbus client and the ability to load
+    the service's active config file.
+    """
+
+    client: DbusClient
+    output: OutputMode
+    _config: Optional[Config] = None
+
+    async def config(self: Self) -> Config:
+        if not self._config:
+            config_file: str = await self.client.config_file
+            config: Config = Config.from_file(config_file)
+            self._config = config
+            return config
+        else:
+            return self._config
+
+
+def pass_config(fn: AsyncCommand) -> AsyncCommand:
+    @click.pass_obj
+    @functools.wraps(fn)
+    async def wrapped(obj: Obj, *args, **kwargs) -> None:
+        config = await obj.config()
+        await fn(config, *args, **kwargs)
+
+    return wrapped
+
+
+def pass_client(fn: AsyncCommand) -> AsyncCommand:
+    """
+    Create a dbus client and pass it to the decorated click handler.
+    """
+
+    @click.pass_obj
+    @functools.wraps(fn)
+    async def wrapped(obj: Obj, *args, **kwargs) -> None:
+        return await fn(obj.client, *args, **kwargs)
+
+    return wrapped
+
+
+@click.group()
+@click.option(
+    "--log-level",
+    envvar="PLUSDECK_LOG_LEVEL",
+    type=click.Choice(["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]),
+    default="INFO",
+    help="Set the log level",
+)
+@click.option(
+    "--output",
+    type=click.Choice(["text", "json"]),
+    default="text",
+    help="Output either human-friendly text or JSON",
+)
+@click.pass_context
+def main(
+    ctx: click.Context,
+    log_level: LogLevel,
+    output: OutputMode,
+) -> None:
+    """
+    Control your Plus Deck 2C Cassette Drive through dbus.
+    """
+
+    logging.basicConfig(level=getattr(logging, log_level))
+
+    # Set the output mode for echo
+    echo.mode = output
+
+    client = DbusClient()
+
+    ctx.obj = Obj(client=client, output=output)
+
+
+@main.group()
+def config() -> None:
+    """
+    Configure plusdeck.
+    """
+    pass
+
+
+@config.command()
+@click.argument("name")
+@async_command
+@pass_config
+async def get(config: Config, name: str) -> None:
+    """
+    Get a parameter from the configuration file.
+    """
+
+    try:
+        echo(config.get(name))
+    except ValueError as exc:
+        echo(str(exc))
+        sys.exit(1)
+
+
+@config.command()
+@async_command
+@pass_config
+async def show(config: Config) -> None:
+    """
+    Show the current configuration.
+    """
+    echo(config)
```

Comment:
A good feature here would be to show drift between the file config and the loaded config. That would involve pulling the config from dbus, presumably.

Reply:
This is implemented, but needs tests.